src/mlpack/methods/edge_boxes/feature_extraction_impl.hpp

```diff
@@ -0,0 +1,1052 @@
+// also stores the locations of max values in Location
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+MaxAndLoc(CubeType& mag, arma::umat& Location, MatType& MaxVal) const
+{
+  /*Vectorize this function after prototype works*/
+  MaxVal = MatType(Location.n_rows, Location.n_cols);
+  for(size_t i = 0; i < mag.n_rows; ++i)
+  {
+    for(size_t j = 0; j < mag.n_cols; ++j)
+    {
+      /*can use -infinity here*/
+      double max =  std::numeric_limits<double>::min();
+      for(size_t k = 0; k < mag.n_slices; ++k)
+      {
+        if(mag(i, j, k) > max)
+        {
+          max = mag(i, j, k);
+          MaxVal(i, j) = max;
+          Location(i, j) = k;
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Computes Gradient, Magnitude & Orientation.
+ */
+template<typename MatType, typename CubeType>
+void StructuredForests<MatType, CubeType>::
+Gradient(const CubeType& InImage, 
+         MatType& Magnitude,
+         MatType& Orientation)
+{
+  const size_t grdNormRad = this->params.GrdNormRad();
+  CubeType dx(InImage.n_rows, InImage.n_cols, InImage.n_slices), 
+             dy(InImage.n_rows, InImage.n_cols, InImage.n_slices);
+
+  dx.zeros();
+  dy.zeros();
+
+  /*
+  From MATLAB documentation:
+  [FX,FY] = gradient(F), where F is a matrix, returns the 
+  x and y components of the two-dImensional numerical gradient. 
+  FX corresponds to ∂F/∂x, the differences in x (horizontal) direction. 
+  FY corresponds to ∂F/∂y, the differences in the y (vertical) direction.
+  */ 
+
+
+  /*
+  gradient calculates the central difference for interior data points.
+  For example, consider a matrix with unit-spaced data, A, that has 
+  horizontal gradient G = gradient(A). The interior gradient values, G(:,j), are:
+
+  G(:,j) = 0.5*(A(:,j+1) - A(:,j-1));
+  where j varies between 2 and N-1, where N is size(A,2).
+
+  The gradient values along the edges of the matrix are calculated with single-sided differences, so that
+
+  G(:,1) = A(:,2) - A(:,1);
+  G(:,N) = A(:,N) - A(:,N-1);
+  
+  The spacing between points in each direction is assumed to be one.
+  */
+  for (size_t i = 0; i < InImage.n_slices; ++i)
+  {
+    dx.slice(i).col(0) = InImage.slice(i).col(1) - InImage.slice(i).col(0);
+    dx.slice(i).col(InImage.n_cols - 1) = InImage.slice(i).col(InImage.n_cols - 1)
+                                        - InImage.slice(i).col(InImage.n_cols - 2);
+
+    for (size_t j = 1; j < InImage.n_cols-1; j++)
+      dx.slice(i).col(j) = 0.5 * ( InImage.slice(i).col(j+1) - InImage.slice(i).col(j) ); 
+
+    // do same for dy.
+    dy.slice(i).row(0) = InImage.slice(i).row(1) - InImage.slice(i).row(0);
+    dy.slice(i).row(InImage.n_rows - 1) = InImage.slice(i).row(InImage.n_rows - 1)
+                                        - InImage.slice(i).row(InImage.n_rows - 2);
+
+    for (size_t j = 1; j < InImage.n_rows-1; j++)
+      dy.slice(i).row(j) = 0.5 * ( InImage.slice(i).row(j+1) - InImage.slice(i).row(j) );
+  }  
+
```

Comment:
Hello @stereomatchingkiss, to save you some time, I made a bunch of similar comments here: https://github.com/mlpack/mlpack/pull/696